### PR TITLE
vLLM-Base: Full enabling of ALiBi

### PR DESCRIPTION
**Pull request by tvoas**
Resolves a regression in ALiBi-based models.

This is a fixed version ofhttps://github.com/HabanaAI/vllm-fork/pull/503 which was reverted by https://github.com/HabanaAI/vllm-fork/pull/907.

This depends on the vllm-hpu-extension PR: https://github.com/HabanaAI/vllm-hpu-extension/pull/141

All pre-commit checks are passed locally.

### CODE REVIEW

**Comment by tannervoas742** [comment]
@michalkuligowski @afierka-intel @madamczyk-intel @kzawora-intel this PR and the extension one https://github.com/HabanaAI/vllm-hpu-extension/pull/141 have been updated and rebased. They picked up the fixes from https://github.com/xhaihao, which resolves all test failures as of ~2.5 weeks ago.

**Comment by adobrzyn** [comment]
Please remove test results from PR comment

**Comment by xuguangxin** [comment]
👏, all tests passed.

File: vllm/attention/backends/hpu_attn.py [hunk 1]
```diff
@@ -510,17 +575,33 @@ def forward(
...
                 **self.common_attention_args(block_list, key_cache,
                                              value_cache))
             output = out.reshape(batch_size, seq_len, hidden_size)
         else:
             # Decoding run.
+            self.position_bias = None
+            alibi_blocks = getattr(attn_metadata, 'alibi_blocks', None)
+            if self.alibi_slopes is not None and alibi_blocks is not None:
+                if (self.prev_attn is not None
+                        and self.prev_attn.tp_rank == self.tp_rank):
```
Comment (madamczyk-intel) [inline vllm/attention/backends/hpu_attn.py:589]: What exactly is tp_rank? Is it releated to tensor parallelism? If so how can it be different between calls?

File: vllm/attention/backends/hpu_attn.py [hunk 2]
```diff
@@ -651,15 +746,54 @@ def _make_alibi_bias(
...
         seq_len,
         padded_len,
         device=alibi_slopes.device,
         dtype=dtype,
-    )[:, :, :, :seq_len].copy_(bias)
-    bias.mul_(alibi_slopes[:, None, None])
-    if num_heads != num_kv_heads:
-        bias = bias.unflatten(1, (num_kv_heads, num_heads // num_kv_heads))
-    return bias
+    )[:, :, :, :seq_len]
```
Comment (madamczyk-intel) [inline vllm/attention/backends/hpu_attn.py:756]: We are we creating a view here? Couldn't we use the whole tensor? Usually there are different problems with views and it's better to avoid them if possible.

File: vllm/attention/backends/hpu_attn.py [hunk 3]
```diff
@@ -425,6 +442,37 @@ def __init__(
...
 
+    def _maybe_init_alibi_biases(
+        self,
+        max_seq_len: int = 4096,
+        prev_attn: Optional[torch.nn.Module] = None,
+    ) -> None:
+        # Set upper bound on sequence length
+        max_seq_len_upper = int(
+            os.getenv(
+                'VLLM_PROMPT_ALIBI_MAX_SEQ_LEN',
```
[the remark this answers is not in the document]
Comment (tvoas) [reply vllm/attention/backends/hpu_attn.py:453]: Max model len is input and output, correct? Here we are only calculating the max sequence length of the input. That said, this code here is likely overly complicated. I will just replace this with ``self.bucketing_ctx.get_max_prompt_shape()``, which is similar.

File: vllm/attention/backends/hpu_attn.py [hunk 4]
```diff
@@ -492,13 +540,30 @@ def forward(
...
+            position_bias = None
+            # If we have alibi_slopes, incorporate them with
+            if (attn_metadata.block_list is None
+                    and self.prompt_position_bias is not None
+                    and self.alibi_slopes is not None):
+                assert attn_bias is not None, \
+                        'attn_bias must be set before calling ' \
+                        'model.forward with alibi biases'
+                slice_1_size = attn_bias.size(-2)
+                slice_2_size = attn_bias.size(-1)
```
[the remark this answers is not in the document]
Comment (tvoas) [reply vllm/attention/backends/hpu_attn.py:552]: I just used the shapes of attn_bias in for extensibility. In practice, I know of no situation where we would not have square biases.

File: vllm/attention/backends/hpu_attn.py [hunk 5]
```diff
@@ -492,13 +540,30 @@ def forward(
...
+                if self.max_seq_len >= max(slice_1_size, slice_2_size):
+                    # Using pre-computed prompt_position_bias subset.
+                    position_bias = self.prompt_position_bias[:, :,
+                                                              -slice_1_size:,
+                                                              -slice_2_size:]
+
+                else:
+                    # For longer sequences than precomputed,
+                    # recreate the bias. This is memory inefficient.
+                    position_bias = _make_prompt_alibi_bias(
```
[the remark this answers is not in the document]
Comment (tvoas) [reply vllm/attention/backends/hpu_attn.py:562]: I believe this is not the correct choice. We could use max_model_len, but these biases are just for prompt phase,  decode phase has biases of a different shape which have much lower memory footprint (a benefit of flat_pa). max_model_len must be large enough to support input + output. If we pre-compute these biases using max_model_len this will waste a large amount of memory in most use cases.

File: vllm/attention/backends/hpu_attn.py (hunk 2 above)
Comment (tvoas) [reply vllm/attention/backends/hpu_attn.py:756]: I will need to profile this. This view was present in the old code, and I assumed the padding was used for a reason. I am not aware why, though, as I didn't look into it.

File: vllm/worker/hpu_model_runner.py [hunk 6]
```diff
@@ -1703,6 +1775,63 @@ def _prepare_decode(
...
+            seq_len = seq_lens[batch_idx]
+            for seq_idx in range(len(block_tables[batch_idx])):
+                block_idx = block_tables[batch_idx][seq_idx]
+
+                # Calculate the number of valid positions in the current block
+                valid_length = seq_len - seq_idx * self.block_size
+                if valid_length > 0:
+                    current_block_length = min(valid_length, self.block_size)
+                    offset_end = current_block_length - valid_length
+                    if offset_end == 0:
```
[the remark this answers is not in the document]
Comment (tvoas) [reply vllm/worker/hpu_model_runner.py:1824]: I believe the slicing could be cleaned up and made more efficient. But these operations are already entirely on the CPU. They get moved to HPU in a single operation later. There should be no concern in terms of graph recompilation, right?

File: vllm/attention/backends/hpu_attn.py (hunk 1 above)
Comment (tvoas) [reply vllm/attention/backends/hpu_attn.py:589]: tp_rank was originally added due to a misunderstanding of how layers are shared when using pipeline parallelism. This check is not actually needed.

**Comment by michalkuligowski** [comment]
/run-gaudi-tests - should now take a little less time due to enabled local package cache just now

File: requirements/hpu.txt [hunk 7]
```diff
@@ -9,4 +9,4 @@ numpy==1.26.4
 tabulate
 setuptools>=77.0.3,<80.0.0
 setuptools-scm>=8
-vllm-hpu-extension @ git+https://github.com/HabanaAI/vllm-hpu-extension.git@f21e935
+vllm-hpu-extension @ git+https://github.com/tannervoas742/vllm-hpu-extension.git@d8d867c
```
Comment (michalkuligowski) [inline requirements/hpu.txt:12]: after related pr in extension is merged please update sha here

**Comment by tvoas** [comment]
@michalkuligowski @madamczyk-intel @kzawora-intel I have rebased the code. Rechecked functional tests (G2 ones) and they passed. Also retested alibi model (Baichuan2) and still functioned with accuracy. Extension PR is also updated.

**Review by madamczyk-intel — APPROVED** [review]
